modules/platforms/dotnet/Apache.Ignite.Core/Impl/Common/TypeCaster.cs

```diff
@@ -87,9 +89,39 @@ private static class Casters<TFrom>
                 }
 
                 var paramExpr = Expression.Parameter(typeof(TFrom));
-                var convertExpr = Expression.Convert(paramExpr, typeof(T));
 
-                return Expression.Lambda<Func<TFrom, T>>(convertExpr, paramExpr).Compile();
+                if (typeof(TFrom) == typeof(Enum))
+                    return Expression.Lambda<Func<TFrom, T>>(TryConvertRawEnum(typeof(T), paramExpr),
```

Comment:
What if `TryConvertRawEnum` returns `null`?

Reply:
Would fail with same conversion error.

Reply:
`Expression.Lambda()` will throw `ArgumentNullException` if `TryConvertRawEnum` returns null. `ArgumentException` is always a bug. Let's check for null and throw `InvalidCastException` explaining that `Enum` can't be converted to the specified type.